src/librustdoc/json/ids.rs

```diff
@@ -0,0 +1,127 @@
+//! Id handling for rustdoc-json.
+//!
+//! Manages the creation of [`rustdoc_json_types::Id`] and the
+//! fact that these don't corespond exactly to [`DefId`], because
+//! [`rustdoc_json_types::Item`] doesn't corespond exactly to what
+//! other phasises thing of as an "item".
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::def::DefKind;
+use rustc_hir::def_id::DefId;
+use rustc_span::{Symbol, sym};
+use rustdoc_json_types as types;
+
+use super::JsonRenderer;
+use crate::clean;
+
+pub(super) type IdInterner = FxHashMap<FullItemId, types::Id>;
+
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+/// An uninterned id.
+///
+/// One of these coresponds to every:
+/// 1. [`rustdoc_json_types::Item`].
+/// 2. [`rustdoc_json_types::Id`] transitivly (as each `Item` has an `Id`).
+///
+/// It's *broadly* equivalent to a [`DefId`], but needs slightly more information
+/// to fully disambiguate items, because sometimes we choose to split a single HIR
+/// item into multiple JSON items, or have items with no coresponding HIR item.
+pub(super) struct FullItemId {
+    /// The "main" id of the item.
+    ///
+    /// In most cases this unequely identifies the item, other fields are just
+    /// used for edge-cases.
+    def_id: DefId,
+
+    /// An extra [`DefId`], which we need for:
+    ///
+    /// 1. Auto-trait impls synthesized by rustdoc.
+    /// 2. Blanket impls synthesized by rustdoc.
+    /// 3. Splitting of reexports of multiple items.
+    ///
+    ///    Eg:
+    ///
+    ///    ```rust
+    ///     mod module {
+    ///         pub struct Foo {} // Exists in type namespace
+    ///         pub fn Foo(){} // Exists in value namespace
+    ///     }
+    ///
+    ///     pub use module::Foo; // Imports both items
+    ///     ```
+    ///
+    ///     In HIR, the `pub use` is just 1 item, but in rustdoc-json it's 2, so
+    ///     we need to disambiguate.
+    extra_id: Option<DefId>,
+
+    /// Needed for `rustc_doc_primitive` modules.
+    ///
+    /// For these, 1 [`DefId`] is used for both the primitive and the fake-module
+    /// that holds it's docs.
+    ///
+    /// N.B. This only matters when documenting the standard library with
+    /// `--document-private-items`. Maybe we should delete that module, and
+    /// remove this.
+    name: Option<Symbol>,
+}
```

Comment:
could this struct be an enum? I'd expect the fields `extra_id` & `name` to never be `Some` simultaneously

Reply:
Also, do we need to store the name? if this only exists to generate 2 rustdoc IDs for what is 1 ID in rustc, this field could be a boolean, say, `is_primitive`, or just an enum variant

Reply:
I guess we could do that, and end up with something like `clean::ItemId` again.

Now I'm almost curious if we could use `clean::ItemId` in place of `json::ids::FullItemId`.

Reply:
Probably not, `FullItemId` would also need a `Primitive` variant then

Reply:
I think it's best to leave that to a followup then. The structure's already sufficiently different.